src/http.c

```diff
@@ -150,75 +155,87 @@ int http_receive(
 		return ERR_HTTP_NO_MEM;
 
-	do {
-		n = safe_ssl_read(tunnel->ssl_handle,
-		                  (uint8_t *) buffer + bytes_read,
-		                  BUFSZ - 1 - bytes_read);
-		if (n > 0) {
-			log_debug_details("%s:\n%s\n", __func__, buffer);
-			const char *eoh;
-
-			bytes_read += n;
-
-			if (!header_size) {
-				/* Did we see the header end? Then get the body size. */
-				eoh = memmem(buffer, bytes_read, "\r\n\r\n", 4);
-				if (eoh) {
-					const char *header;
-
-					header = find_header(
-					                 buffer,
-					                 "Content-Length: ", BUFSZ);
-					header_size = eoh - buffer + 4;
-					if (header)
-						content_size = atoi(header);
-
-					if (find_header(buffer,
-					                "Transfer-Encoding: chunked",
-					                BUFSZ))
-						chunked = 1;
-				}
+	while (1) {
+		int n;
+
+		while ((n = safe_ssl_read(tunnel->ssl_handle,
+		                          (uint8_t *) buffer + bytes_read,
+		                          capacity - bytes_read)) == ERR_SSL_AGAIN)
+			;
+		if (n < 0) {
+			log_error("Error reading from SSL connection (%s).\n",
+			          err_ssl_str(n));
+			free(buffer);
+			return ERR_HTTP_SSL;
+		}
+		bytes_read += n;
+
+		log_debug_details("%s:\n%s\n", __func__, buffer);
+
+		if (!header_size) {
+			/* Have we reached the end of the HTTP header? */
+			static const char EOH[4] = "\r\n\r\n";
+			const char *eoh = memmem(buffer, bytes_read,
+			                         EOH, sizeof(EOH));
+
+			if (eoh) {
+				header_size = eoh - buffer + sizeof(EOH);
+
+				/* Get the body size. */
+				const char *header = find_header(buffer,
+				                                 "Content-Length: ",
+				                                 header_size);
+
+				if (header)
+					content_size = atoi(header);
+
+				if (find_header(buffer,
+				                "Transfer-Encoding: chunked",
+				                header_size))
+					chunked = 1;
 			}
+		}
 
-			if (header_size) {
-				/* We saw the whole header, is the body done as well? */
-				if (chunked) {
-					/* Last chunk terminator. Done naively. */
-					if (bytes_read >= 7 &&
-					    !memcmp(&buffer[bytes_read - 7],
-					            "\r\n0\r\n\r\n", 7))
-						break;
-				} else {
-					if (bytes_read >= header_size + content_size)
-						break;
-				}
+		if (header_size) {
+			/* Have we reached the end of the HTTP body? */
+			if (chunked) {
+				static const char EOB[7] = "\r\n0\r\n\r\n";
+
+				/* Last chunk terminator. Done naively. */
+				if (bytes_read >= sizeof(EOB) &&
+				    !memcmp(&buffer[bytes_read - sizeof(EOB)],
+				            EOB, sizeof(EOB)))
+					break;
+			} else {
+				if (bytes_read >= header_size + content_size)
+					break;
 			}
+		}
 
-			if (bytes_read == BUFSZ - 1) {
-				log_warn("Response too big\n");
+		/* expand the buffer if necessary */
+		if (bytes_read == capacity) {
+			char *new_buffer;
+
+			assert(UINT32_MAX / capacity >= 2);
+			capacity *= 2;
+			new_buffer = realloc(buffer, capacity);
```

Comment:
...maybe you have decided to read to the very end of the buffer because you reallocate it here?

Reply:
That's not related - I think.

The only difference is that we used to bail out with `HTTP_ERR_NO_MEM`. Now we increase the size of the buffer.

Reply:
Besides I don't think I am now reading more than previously, unless of course we were hitting `ERR_HTTP_TOO_LONG` which is always a bug (see #575 for example). On the contrary I will read less and bail out in case of an SSL violation.

Reply:
I see.